File: poc-csharp/RpsPoc/Ui/MatchView.cs
using Avalonia;
using Avalonia.Controls;
using Avalonia.Layout;
using Avalonia.Media;

using RpsPoc.Domain;

namespace RpsPoc.Ui;

/// <summary>
/// Um match desenhado: dois retangulos coloridos pelo score de cada jogador,
/// com a jogada corrente no meio. So le estado, nao decide nada.
/// </summary>
public sealed class MatchView : Grid
{
    private readonly Border _left = new() { CornerRadius = new CornerRadius(6) };
    private readonly Border _right = new() { CornerRadius = new CornerRadius(6) };
    private readonly TextBlock _leftMove = new() { Text = "...", FontSize = 11 };
    private readonly TextBlock _rightMove = new() { Text = "...", FontSize = 11 };
    private readonly TextBlock _title = new() { FontSize = 12, Foreground = Brushes.Gainsboro };

    public MatchView(Match match, int maxScore)
    {
        Match = match;
        Margin = new Thickness(4);
        RowDefinitions = new RowDefinitions("Auto,*");

        _title.Text = $"match #{match.Id}";
        SetRow(_title, 0);
        Children.Add(_title);

        var arena = new Grid { ColumnDefinitions = new ColumnDefinitions("*,*") };
        Fill(_left, match.P1, _leftMove);
        Fill(_right, match.P2, _rightMove);
        SetColumn(_left, 0);
        SetColumn(_right, 1);
        arena.Children.Add(_left);
        arena.Children.Add(_right);

        SetRow(arena, 1);
        Children.Add(arena);

        Refresh(maxScore);
    }

    public Match Match { get; }

    /// <summary>Repinta a partir do estado atual do match. Sempre na UI thread.</summary>
    public void Refresh(int maxScore)
    {
        _left.Background = ColorMapper.ScoreToBrush(Match.P1.Score, maxScore);
        _right.Background = ColorMapper.ScoreToBrush(Match.P2.Score, maxScore);

        var round = Match.LastRound;
        _leftMove.Text = round is null ? "..." : round.P1.ToString();
        _rightMove.Text = round is null ? "..." : round.P2.ToString();

        if (round is null)
        {
            return;
        }

        if (round.IsDraw)
        {
            _title.Text = $"match #{Match.Id} — empate, replay";
            _left.Opacity = 1;
            _right.Opacity = 1;
            return;
        }

        // Rodada decisiva: apaga o perdedor. O runner ainda segura o match na
        // arena por um instante antes de publicar MatchEnded.
        var p1Won = round.P1.Beats(round.P2) > 0;
        _title.Text = $"match #{Match.Id} — vence {(p1Won ? Match.P1 : Match.P2)}";
        (p1Won ? _right : _left).Opacity = 0.25;
    }

    private static void Fill(Border border, Player player, TextBlock moveText)
    {
        border.Margin = new Thickness(3);
        border.Child = new StackPanel
        {
            HorizontalAlignment = HorizontalAlignment.Center,
            VerticalAlignment = VerticalAlignment.Center,
            Spacing = 2,
            Children =
            {
                new TextBlock
                {
                    Text = player.ToString(),
                    FontSize = 13,
                    Foreground = Brushes.Black,
                    HorizontalAlignment = HorizontalAlignment.Center,
                },
                moveText,
            },
        };
        moveText.Foreground = Brushes.Black;
        moveText.HorizontalAlignment = HorizontalAlignment.Center;
    }
}

File: poc-java/src/main/java/rps/ui/MatchView.java
package rps.ui;

import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.scene.control.Label;
import javafx.scene.layout.Background;
import javafx.scene.layout.BackgroundFill;
import javafx.scene.layout.CornerRadii;
import javafx.scene.layout.HBox;
import javafx.scene.layout.Priority;
import javafx.scene.layout.VBox;
import javafx.scene.paint.Color;
import javafx.scene.text.Font;

import rps.domain.Match;
import rps.domain.Round;

/**
 * Um match desenhado: dois retangulos coloridos pelo score de cada jogador,
 * com a jogada corrente no meio. So le estado, nao decide nada.
 */
public final class MatchView extends VBox {

    private final Match match;
    private final VBox left = new VBox();
    private final VBox right = new VBox();
    private final Label leftMove = new Label("...");
    private final Label rightMove = new Label("...");
    private final Label title = new Label();

    public MatchView(Match match, int maxScore) {
        this.match = match;
        setSpacing(4);
        setPadding(new Insets(6));

        title.setFont(Font.font(12));
        title.setTextFill(Color.web("#cfd3dc"));
        title.setText("match #" + match.id());

        HBox arena = new HBox(6);
        HBox.setHgrow(left, Priority.ALWAYS);
        HBox.setHgrow(right, Priority.ALWAYS);
        VBox.setVgrow(arena, Priority.ALWAYS);
        arena.getChildren().addAll(side(left, match.p1(), leftMove), side(right, match.p2(), rightMove));

        getChildren().addAll(title, arena);
        refresh(maxScore);
    }

    private VBox side(VBox box, rps.domain.Player player, Label moveLabel) {
        box.setAlignment(Pos.CENTER);
        box.setSpacing(2);
        Label name = new Label(player.toString());
        name.setFont(Font.font(13));
        name.setTextFill(Color.web("#101318"));
        moveLabel.setFont(Font.font(11));
        moveLabel.setTextFill(Color.web("#101318"));
        box.getChildren().addAll(name, moveLabel);
        HBox.setHgrow(box, Priority.ALWAYS);
        box.setMaxWidth(Double.MAX_VALUE);
        box.setMaxHeight(Double.MAX_VALUE);
        return box;
    }

    /** Repinta a partir do estado atual do match. Chamado sempre na UI thread. */
    public void refresh(int maxScore) {
        paint(left, match.p1().score(), maxScore);
        paint(right, match.p2().score(), maxScore);

        Round round = match.lastRound();
        leftMove.setText(round == null ? "..." : round.p1().name());
        rightMove.setText(round == null ? "..." : round.p2().name());

        if (round == null) {
            return;
        }
        if (round.isDraw()) {
            title.setText("match #" + match.id() + " — empate, replay");
            left.setOpacity(1);
            right.setOpacity(1);
            return;
        }
        // Rodada decisiva: apaga o perdedor. O runner ainda vai segurar o match
        // na arena por um instante antes de publicar MatchEnded.
        boolean p1Won = round.p1().beats(round.p2()) > 0;
        title.setText("match #" + match.id() + " — vence " + (p1Won ? match.p1() : match.p2()));
        (p1Won ? right : left).setOpacity(0.25);
    }

    private void paint(VBox box, int score, int maxScore) {
        Color color = ColorMapper.scoreToColor(score, maxScore);
        box.setBackground(new Background(new BackgroundFill(color, new CornerRadii(6), Insets.EMPTY)));
    }

    public Match match() {
        return match;
    }
}
